### ISSUE TITLE
Feature request: Extensibility of the standalone Quarto visual editor

### ISSUE THREAD

**Issue by gvelasq** [quarto-dev/quarto-cli#388]
It's been great to experiment with RStudio's transition to [Electron](https://dailies.rstudio.com/rstudio/spotted-wakerobin/electron/bionic/). Will the forthcoming [standalone Quarto visual editor](https://quarto.org/docs/visual-editor/#getting-started) also be based on Electron?

If so, given the runaway success of extension registries for Electron-based text editors (e.g. [Obsidian](https://obsidian.md/plugins), [VS Code](https://marketplace.visualstudio.com/vscode), [OpenVSX](https://open-vsx.org/)), do you anticipate making an extension API (e.g., [Obsidian API](https://github.com/obsidianmd/obsidian-api), [VS Code API](https://code.visualstudio.com/api)) available for the standalone Quarto visual editor?

A final thought is, given the polyglot spirit of Quarto, would such an API have bindings to JavaScript/TypeScript only, or could it have bindings to R and Python?

**Comment by jjallaire** [comment]
Yes, it will be based on Electron and there will definitely be an extension API. The full fidelity API will likely use JS/TS. It would be possible to create limited bindings to R/Python, but we also don't want to create an extension ecosystem where end users need to install a runtime + a bunch of packages for a language they don't even use just to get an extension working. The best path to writing extensions in other languages will likely be WASM.

**Comment by gvelasq** [comment]
Thanks very much for the response. This is exciting news on all counts. It will be great to see what the community can build with the JS/TS extensions API for the standalone Quarto visual editor.

I appreciate that full-fidelity API support may not be possible for R/Python, especially if it would require additional complexities for end users. On the developer side, perhaps there will be solutions as you mentioned that could expand the pool of potential extension contributors to other languages, without requiring the complexities of the 'glue' to be passed onto end users.

Another thought that comes to mind is that rstudioapi, RStudio Addins, Shiny Gadgets, rsthemes, etc. are already a great way for R users to extend RStudio, some of which have already been ported [outside](https://github.com/REditorSupport/vscode-R/wiki/RStudio-addin-support) of RStudio. Perhaps an abstraction of addins as language-specific extension mini-ecosystems for R, Python, Julia, etc. might coexist with each other and with the formal JS/TS extensions API in the standalone Quarto editor, since users conversant in a particular language will opt-in to their language-specific runtimes/packages.

**Comment by jjallaire** [comment]
Yes, I agree that some extensions will absolutely required and benefit from the use of R or Python (as w/ the examples you cite). So there will definitely need to be both available in some form!
